packages/oidc/src/token.ts

```diff
@@ -0,0 +1,226 @@
+  if (isBrowserEnv()) {
+    return '';
+  }
+  const vercelFolder = 'com.vercel.cli';
+  const os = await import('os');
+  const path = await import('path');
+  if (process.env.XDG_DATA_HOME) {
+    return path.join(process.env.XDG_DATA_HOME, vercelFolder);
+  }
+  switch (os.platform()) {
+    case 'darwin':
+      return path.join('~/Library/Application Support', vercelFolder);
+    case 'linux':
+      return path.join('~/.local/share', vercelFolder);
+    case 'win32':
+      return path.join('%LOCALAPPDATA%', vercelFolder);
+    default:
+      return null;
+  }
+}
+
+async function getVercelCliToken(): Promise<string | null> {
+  const dataDir = await getVercelDataDir();
+  if (!dataDir) {
+    return null;
+  }
+  const path = await import('path');
+  const fs = await import('fs');
+  const tokenPath = path.join(dataDir, 'auth.json');
+  if (!fs.existsSync(tokenPath)) {
+    return null;
+  }
+  const token = fs.readFileSync(tokenPath, 'utf8');
+  return JSON.parse(token).token;
+}
+
+interface VercelTokenResponse {
+  token: string;
+}
+
+async function getVercelOidcToken(
+  authToken: string,
+  projectId: string
+): Promise<VercelTokenResponse | null> {
+  try {
+    const url = `https://api.vercel.com/projects/token?projectId=${projectId}&source=vercel-oidc-refresh`;
+    const res = await fetch(url, {
+      method: 'POST',
+      headers: {
+        Authorization: `Bearer ${authToken}`,
+      },
+    });
+    if (!res.ok) {
+      throw new Error(`failed to refresh token: ${res.statusText}`);
+    }
+    if (!res.json) {
+      throw new Error('failed to refresh token: no json response');
+    }
+    const tokenRes = (await res.json()) as VercelTokenResponse;
+    return tokenRes;
+  } catch (e) {
+    const message = e instanceof Error ? e.message : e;
+    throw new Error(`failed to refresh token: ${message}`);
+  }
+}
+
+async function findProjectId(): Promise<string> {
+  const dir = await findRootDir();
+  if (!dir) {
+    throw new Error('unable to find root directory');
+  }
+  try {
+    const fs = await import('fs');
+    const path = await import('path');
+    const pkgPath = path.join(dir, '.vercel', 'project.json');
+    if (!fs.existsSync(pkgPath)) {
+      throw new Error('project.json not found');
+    }
+    const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf8'));
+    return pkg.projectId;
+  } catch (e) {
+    const message = e instanceof Error ? e.message : e;
+    throw new Error(`unable to find project id: ${message}`);
+  }
+}
+
+async function findRootDir(): Promise<string> {
+  try {
+    const fs = await import('fs');
+    const path = await import('path');
+    let dir = process.cwd();
+    while (dir !== path.dirname(dir)) {
+      const pkgPath = path.join(dir, '.vercel');
+      if (fs.existsSync(pkgPath)) {
+        return dir;
+      }
+      dir = path.dirname(dir);
+    }
+  } catch (e) {
+    throw new Error('token refresh only supported in node server environments');
```

Comment:
Style question: are we typically doing lowercase error messages in `@vercel/` packages, including CLI? I would guess we should do sentence-case and optionally punctuate

Reply:
it's different in literally every programming language I've used, I'll check what's going on in the rest of the project and match it lol